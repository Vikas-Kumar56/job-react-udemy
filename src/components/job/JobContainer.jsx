import React, { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Button, Box, styled } from '@material-ui/core';
import { selectAllJobs, fetchJobs } from '../../features/job/jobSlice';
import JobList from './JobList';
import JobLoadingSkeleton from './JobLoadingSkeleton';

const StyledBox = styled(Box)({
  display: 'flex',
  justifyContent: 'center',
  flexDirection: 'row',
  marginTop: '4em',
});

const StyledButton = styled(Button)({
  width: '20%',
  height: '4em',
});

const JobContainer = () => {
  const [filter, setFilter] = useState({ limit: 10, offset: 0 });

  const dispatch = useDispatch();
  const jobs = useSelector(selectAllJobs);

  const jobStatus = useSelector((state) => state.jobs.status);

  useEffect(() => {
    dispatch(fetchJobs(filter));
  }, [dispatch, filter]);

  const fetchMoreJobs = () => {
    const { limit, offset } = filter;
    setFilter({ limit, offset: limit + offset });
  };

  const getView = () => {
    if (jobStatus === 'failed') {
      return <div>Something went wrong!</div>;
    }

    return (
      <>
        <JobList jobs={jobs} />
        {jobStatus === 'loading' && <JobLoadingSkeleton />}
        {jobs.length !== 0 && (
          <StyledBox>
            <StyledButton
              data-testid='load-more-btn'
              variant='contained'
              color='primary'
              disabled={jobStatus === 'loading'}
              onClick={fetchMoreJobs}
            >
              Load More
            </StyledButton>
          </StyledBox>
        )}
      </>
    );
  };

  return <>{getView()}</>;
};

export default JobContainer;
