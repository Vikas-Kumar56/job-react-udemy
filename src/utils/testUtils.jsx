import React from 'react';
import { render } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';

// import here all your reducer
import jobReducer from '../features/job/jobSlice';

function renderWithRedux(
  ui,
  {
    preloadedState,
    store = configureStore({ reducer: { jobs: jobReducer }, preloadedState }),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }

  return render(ui, { wrapper: Wrapper, ...renderOptions });
}

export * from '@testing-library/react';

export { renderWithRedux };
