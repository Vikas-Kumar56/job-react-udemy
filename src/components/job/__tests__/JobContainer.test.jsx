import React from 'react';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import { renderWithRedux, fireEvent } from '../../../utils/testUtils';
import JobContainer from '../JobContainer';

const handlers = [
  rest.get('http://localhost:5000/api/v1/jobs', (req, res, ctx) => {
    const offset = req.url.searchParams.get('offset');
    if (+offset === 0) {
      return res(
        ctx.json({
          jobs: [
            {
              createdAt: '14/08/2021',
              description: 'description',
              expiredAt: '2022-10-10T00:00:00.000Z',
              id: 'b8d86ffc-da29-4bf5-9d72-f4bd1b76d89a',
              maxBudget: '200',
              minBudget: '100',
              skills: 'skills',
              title: 'title test',
              updatedAt: '14/08/2021',
              userId: 'ef3a51a3-642a-4230-9a01-ecd475e72f07',
              version: 'ede2bf46-d06a-4d1f-b756-b718de36165b',
            },
          ],
        }),
        ctx.delay(150)
      );
    }
    return res(
      ctx.json({
        jobs: [
          {
            createdAt: '14/08/2021',
            description: 'description - 2',
            expiredAt: '2022-10-10T00:00:00.000Z',
            id: 'b8d86ffc-da29-4bf5-9d72-f4bd1b76d89b',
            maxBudget: '200',
            minBudget: '100',
            skills: 'skills',
            title: 'title test - 2',
            updatedAt: '14/08/2021',
            userId: 'ef3a51a3-642a-4230-9a01-ecd475e72f07',
            version: 'ede2bf46-d06a-4d1f-b756-b718de36165b',
          },
        ],
      }),
      ctx.delay(150)
    );
  }),
];

const server = setupServer(...handlers);

describe('JobContainer', () => {
  beforeAll(() => server.listen());

  afterEach(() => server.resetHandlers());

  afterAll(() => server.close());

  it('should show loader when request is pending', () => {
    const { getByTestId, asFragment } = renderWithRedux(<JobContainer />);

    expect(getByTestId('job-container-loading')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render jobList', async () => {
    const { findByText, asFragment } = renderWithRedux(<JobContainer />);

    expect(await findByText('title test')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('should load more jobs when click on load more button', async () => {
    const { findByText, getByText, asFragment, getByTestId, queryByTestId } =
      renderWithRedux(<JobContainer />);
    expect(await findByText('title test')).toBeInTheDocument();

    fireEvent.click(getByText('Load More'));
    expect(getByTestId('load-more-btn')).toBeDisabled();
    expect(getByTestId('job-container-loading')).toBeInTheDocument();

    expect(await findByText('title test - 2')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();

    expect(getByTestId('load-more-btn')).not.toBeDisabled();
    expect(queryByTestId('job-container-loading')).not.toBeInTheDocument();
  });

  it('should handle server error', async () => {
    server.use(
      rest.get('http://localhost:5000/api/v1/jobs', (req, res, ctx) => {
        return res.once(
          ctx.status(500),
          ctx.json({ mesage: 'server is not working!' })
        );
      })
    );

    const { findByText } = renderWithRedux(<JobContainer />);

    expect(await findByText('Something went wrong!')).toBeInTheDocument();
  });
});
