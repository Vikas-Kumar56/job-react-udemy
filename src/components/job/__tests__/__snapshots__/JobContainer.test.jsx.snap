// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobContainer should load more jobs when click on load more button 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="MuiPaper-root WithStyles(ForwardRef(Paper))-root-14 MuiPaper-elevation2 MuiPaper-rounded"
      >
        <div
          class="MuiGrid-root MuiGrid-container"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-9"
          >
            <h1
              class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
            >
              title test
              <span
                class="MuiBadge-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root memo-root-15"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 1L3 5v6c0 5.55 3.84 10.74 9 12 5.16-1.26 9-6.45 9-12V5l-9-4zm-2 16l-4-4 1.41-1.41L10 14.17l6.59-6.59L18 9l-8 8z"
                  />
                </svg>
                <span
                  class="MuiBadge-badge MuiBadge-anchorOriginTopRightRectangle MuiBadge-invisible"
                />
              </span>
            </h1>
            <p
              class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom"
            >
              description
            </p>
            <p
              class="MuiTypography-root WithStyles(ForwardRef(Typography))-root-16 MuiTypography-body1 MuiTypography-gutterBottom"
            >
              skills
            </p>
            <span
              class="WithStyles(ForwardRef(Typography))-root-17"
              variant="body1"
            >
              Posted Date: 14/08/2021
            </span>
            <span
              class="WithStyles(ForwardRef(Typography))-root-17"
              variant="body1"
            >
              Expired Date: 2022-10-10T00:00:00.000Z
            </span>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
          >
            <div
              class="MuiBox-root MuiBox-root-18"
            >
              <p
                class="MuiTypography-root WithStyles(ForwardRef(Typography))-root-19 MuiTypography-body1"
              >
                $100 - $200
              </p>
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-text WithStyles(ForwardRef(Button))-root-20 MuiButton-textSecondary"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-label"
                >
                  Place Bid
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="MuiPaper-root WithStyles(ForwardRef(Paper))-root-14 MuiPaper-elevation2 MuiPaper-rounded"
      >
        <div
          class="MuiGrid-root MuiGrid-container"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-9"
          >
            <h1
              class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
            >
              title test - 2
              <span
                class="MuiBadge-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root memo-root-15"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 1L3 5v6c0 5.55 3.84 10.74 9 12 5.16-1.26 9-6.45 9-12V5l-9-4zm-2 16l-4-4 1.41-1.41L10 14.17l6.59-6.59L18 9l-8 8z"
                  />
                </svg>
                <span
                  class="MuiBadge-badge MuiBadge-anchorOriginTopRightRectangle MuiBadge-invisible"
                />
              </span>
            </h1>
            <p
              class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom"
            >
              description - 2
            </p>
            <p
              class="MuiTypography-root WithStyles(ForwardRef(Typography))-root-16 MuiTypography-body1 MuiTypography-gutterBottom"
            >
              skills
            </p>
            <span
              class="WithStyles(ForwardRef(Typography))-root-17"
              variant="body1"
            >
              Posted Date: 14/08/2021
            </span>
            <span
              class="WithStyles(ForwardRef(Typography))-root-17"
              variant="body1"
            >
              Expired Date: 2022-10-10T00:00:00.000Z
            </span>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
          >
            <div
              class="MuiBox-root MuiBox-root-25"
            >
              <p
                class="MuiTypography-root WithStyles(ForwardRef(Typography))-root-19 MuiTypography-body1"
              >
                $100 - $200
              </p>
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-text WithStyles(ForwardRef(Button))-root-20 MuiButton-textSecondary"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-label"
                >
                  Place Bid
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="MuiBox-root MuiBox-root-22 Styled(MuiBox)-root-21"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained WithStyles(ForwardRef(Button))-root-23 MuiButton-containedPrimary"
      data-testid="load-more-btn"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Load More
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`JobContainer should render jobList 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="MuiPaper-root WithStyles(ForwardRef(Paper))-root-3 MuiPaper-elevation2 MuiPaper-rounded"
      >
        <div
          class="MuiGrid-root MuiGrid-container"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-9"
          >
            <h1
              class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
            >
              title test
              <span
                class="MuiBadge-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root memo-root-4"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 1L3 5v6c0 5.55 3.84 10.74 9 12 5.16-1.26 9-6.45 9-12V5l-9-4zm-2 16l-4-4 1.41-1.41L10 14.17l6.59-6.59L18 9l-8 8z"
                  />
                </svg>
                <span
                  class="MuiBadge-badge MuiBadge-anchorOriginTopRightRectangle MuiBadge-invisible"
                />
              </span>
            </h1>
            <p
              class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom"
            >
              description
            </p>
            <p
              class="MuiTypography-root WithStyles(ForwardRef(Typography))-root-5 MuiTypography-body1 MuiTypography-gutterBottom"
            >
              skills
            </p>
            <span
              class="WithStyles(ForwardRef(Typography))-root-6"
              variant="body1"
            >
              Posted Date: 14/08/2021
            </span>
            <span
              class="WithStyles(ForwardRef(Typography))-root-6"
              variant="body1"
            >
              Expired Date: 2022-10-10T00:00:00.000Z
            </span>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
          >
            <div
              class="MuiBox-root MuiBox-root-7"
            >
              <p
                class="MuiTypography-root WithStyles(ForwardRef(Typography))-root-8 MuiTypography-body1"
              >
                $100 - $200
              </p>
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-text WithStyles(ForwardRef(Button))-root-9 MuiButton-textSecondary"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-label"
                >
                  Place Bid
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="MuiBox-root MuiBox-root-11 Styled(MuiBox)-root-10"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained WithStyles(ForwardRef(Button))-root-12 MuiButton-containedPrimary"
      data-testid="load-more-btn"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Load More
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`JobContainer should show loader when request is pending 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
  />
  <div
    class="MuiPaper-root WithStyles(ForwardRef(Paper))-root-1 MuiPaper-elevation1 MuiPaper-rounded"
    data-testid="job-container-loading"
  >
    <div
      class="MuiGrid-root MuiGrid-container"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-9"
      >
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse"
          style="height: 30px;"
        />
        <span
          class="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-pulse"
          style="height: 100px;"
        />
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse"
          style="height: 30px;"
        />
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1"
      />
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
      >
        <span
          class="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-pulse"
          style="height: 160px;"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
